### ISSUE TITLE
Filtering not available on ThenInclude

### ISSUE THREAD

**Issue by cjheath** [dotnet/EntityFramework.Docs#3964]
Include() supports filtering, but not ThenInclude. This seems like an odd oversight.
```
    var people = ctx.People
	    .Where(person => !person.IsDeleted)
	    .Include(person => person.CardAssignments
		.Where(ca => ca.DeassignDateTime == null)    // This is ok, filter on Include
	    )
	    .ThenInclude(ca => ca.Card
	        .Where(c => !c.IsDeleted)          // This is not supported, filter on ThenInclude
	    );
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 25a3ce6e-43ab-e0af-01b3-a82fdaf3cb31
* Version Independent ID: 320e7305-e0a3-a1c6-77f2-1d3029b660b1
* Content: [Eager Loading of Related Data - EF Core](https://docs.microsoft.com/en-us/ef/core/querying/related-data/eager)
* Content Source: [entity-framework/core/querying/related-data/eager.md](https://github.com/dotnet/EntityFramework.Docs/blob/main/entity-framework/core/querying/related-data/eager.md)
* Product: **entity-framework**
* Technology: **entity-framework-core**
* GitHub Login: @roji
* Microsoft Alias: **avickers**

**Comment by smitpatel** [comment]
`ca.Card` is a reference navigation. Filtering is not supported on reference navigations. It can only be applied on collection navigations. It works for Include/ThenInclude both.

**Comment by cjheath** [comment]
Ugh, right, that makes sense.

However, `ca.Card` is non-nullable, so filtering on it should exclude the CardAssignment from the parent list.
I recognise that this is a significant extension of the current design intention, but it is totally a useful thing to have.
Perhaps this should be considered an enhancement request?

**Comment by smitpatel** [comment]
Elaborate more or perhaps provide runnable code on what you are trying to accomplish.

**Comment by cjheath** [comment]
Original was a bad example, and I've solved he original problem. However, this kind of problem does occur, and since you asked for an example, here's a simplified card-game schema to formulate the example:
```
CREATE TABLE Card (
	-- Card has Suit that has Suit Suit Name
	SuitName                                VARCHAR NOT NULL,
	-- Card has Rank
	[Rank]                                  VARCHAR NOT NULL CHECK([Rank] = '10' OR [Rank] = '2' OR [Rank] = '3' OR [Rank] = '4' OR [Rank] = '5' OR [Rank] = '6' OR [Rank] = '7' OR [Rank] = '8' OR [Rank] = '9' OR [Rank] = 'Ace' OR [Rank] = 'Jack' OR [Rank] = 'Joker' OR [Rank] = 'King' OR [Rank] = 'Queen'),
	-- Primary index to Card(Suit, Rank in "Card has Suit", "Card has Rank")
	PRIMARY KEY CLUSTERED(SuitName, [Rank])
)
GO


CREATE TABLE Holding (
	-- Holding involves Player that has Player ID
	PlayerID                                BIGINT NOT NULL,
	-- Holding involves Card that has Suit that has Suit Suit Name
	CardSuitName                            VARCHAR NOT NULL CHECK(CardSuitName = 'Clubs' OR CardSuitName = 'Diamonds' OR CardSuitName = 'Hearts' OR CardSuitName = 'Spades'),
	-- Holding involves Card that has Rank
	CardRank                                VARCHAR NOT NULL,
	-- Primary index to Holding(Player, Card in "Player holds Card")
	PRIMARY KEY CLUSTERED(PlayerID, CardSuitName, CardRank),
	FOREIGN KEY (CardSuitName, CardRank) REFERENCES Card (SuitName, [Rank])
)
GO


CREATE TABLE Player (
	-- Player has Player ID
	PlayerID                                BIGINT NOT NULL IDENTITY,
	-- Primary index to Player(Player ID in "Player has Player ID")
	PRIMARY KEY CLUSTERED(PlayerID)
)
GO


CREATE TABLE Suit (
	-- Suit has Suit Suit Name
	SuitName                                VARCHAR NOT NULL CHECK(SuitName = 'Clubs' OR SuitName = 'Diamonds' OR SuitName = 'Hearts' OR SuitName = 'Spades'),
	-- Suit has Colour
	Colour                                  VARCHAR NOT NULL CHECK(Colour = 'Black' OR Colour = 'Red'),
	-- Primary index to Suit(Suit Suit Name in "Suit has Suit Suit Name")
	PRIMARY KEY CLUSTERED(SuitName)
)
GO


ALTER TABLE Card
	ADD FOREIGN KEY (SuitName) REFERENCES Suit (SuitName)
GO

ALTER TABLE Holding
	ADD FOREIGN KEY (PlayerID) REFERENCES Player (PlayerID)
GO
```

I might want to know which players are holding Hearts:
```
var players = ctx.Players
    .Include(player => player.Holdings)
    .ThenInclude(holding => holding.Suit
        .Where(suit => suit.Name == 'Hearts'
    )
```

Every `Holding` is a single card of just one suit, but I only want to know about the Hearts. Or perhaps, only the Red cards. Etc...

**Comment by cjheath** [comment]
FWIW, the original (real) case for this emerged. It's a multi-tenanted database. Person and Card tables both contain a TenantID, but CardAssignment doesn't. CardAssignment doesn't necessarily contain a Person. All CardAssignments must be loaded for a specific Tenant. So I actually want to say something like:

```
var ca = ctx.CardAssignments
	    .Include(ca => ca.Card
                .Where(card => card.TenantID = tenant.Id)
            )
	    .Include(ca => ca.Person)
```

Note that the CardAssignment should not be loaded unless the Card in its reference navigation is for the correct Tenant. So this is a slightly different case than in my previous example, but it represents the same need - to filter a set based on the condition of a reference object.

**Comment by smitpatel** [comment]
Few observations
- Query is invalid. Player has holdings, holding has card, card has suit. Holding.Suit doesn't exist.
- When you say you want to know which players are holding hearts, then why the query is doing Include. Include is about loading related data. There is no data to load if you just want to know players.
- If you have tenantId filter then it is more appropriate to use global query filters.

I am still unsure what is the query intention here to suggest way to write.

**Comment by cjheath** [comment]
It seems I mixed up two versions of my card player example, or the scaffolding found it didn't need to create an explicit Holding model (it does, but I deleted the columns which require that). Regardless of my errors, I don't understand your difficulty in following my intention here. Your response seems a little too defensive, to be honest.

It's not complicated. I want all players, and for each player, I want all the cards each is holding - but only if that card is a Heart.

Yes, there may be different ways to massage this simple example. Not so in more complicated cases.
BTW, I've written my own query language (in the past) and this kind of query is common. But EFcore doesn't handle it.
